tsdb/engine/tsm1/compact.go

```diff
@@ -1128,33 +1134,35 @@ func (c *Compactor) write(path string, iter KeyIterator, throttle bool, logger *
 	// in memory.
 	if iter.EstimatedIndexSize() > 64*1024*1024 {
 		w, err = NewTSMWriterWithDiskBuffer(limitWriter)
-		if err != nil {
-			return err
-		}
 	} else {
 		w, err = NewTSMWriter(limitWriter)
-		if err != nil {
-			return err
-		}
 	}
-
+	if err != nil {
+		// Close the file and return if we can't create the TSMWriter
+		return false, errors.Join(err, fd.Close())
+	}
 	defer func() {
-		closeErr := w.Close()
-		if err == nil {
-			err = closeErr
-		}
+		var eInProgress errCompactionInProgress
 
-		// Check for errors where we should not remove the file
-		_, inProgress := err.(errCompactionInProgress)
-		maxBlocks := err == ErrMaxBlocksExceeded
-		maxFileSize := err == errMaxFileExceeded
-		if inProgress || maxBlocks || maxFileSize {
+		errs := make([]error, 0, 3)
+		errs = append(errs, err)
+		closeErr := w.Close()
+		errs = append(errs, closeErr)
+
+		// Check for conditions where we should not remove the file
+		inProgress := errors.As(err, &eInProgress)
+		inProgress = inProgress && errors.Is(eInProgress.err, os.ErrExist)
+		maxBlocks := errors.Is(err, ErrMaxBlocksExceeded)
+		maxFileSize := errors.Is(err, errMaxFileExceeded)
+		if (closeErr == nil) && (inProgress || maxBlocks || maxFileSize) {
+			// do not join errors, there is only the one.
+			// rollToNext was set in the main body of the method
```

Comment:
Since we don't return `closeErr` in this case, is there a way we can log it?

Also, would it make more sense to check if `rollToNext` is set instead of checking for the conditions where `rollToNext` would be set? This feels like something that could get out of sync with the main function.

Reply:
`closeErr` is always nil in this branch of the conditional. In the other branch, it is joined with any other errors and returned.

Good idea on checking `rollToNext` here. I will change to do that.